src/proposal.mligo

```diff
@@ -337,6 +382,40 @@ let delete_proposal
     Set.remove (start_date, proposal_key) store.proposal_key_list_sort_by_date
   }
 
+let update_quorum(store, config : storage * config): storage =
+  let current_period = get_current_period_num(store.start_time, config.voting_period) in
+  if store.quorum_threshold_at_stage.last_updated_period = current_period
+    then store // Quorum has been updated in this period, so no change is required.
+    else
+      let last_period_frozen_tokens = match Map.find_opt store.frozen_token_id store.total_supply with
+        | Some (n) -> n
+            // We assume that this function will always be called before freezing
+            // tokens, so that if we are in this branch (where quorum_threshold_at_stage
+            // is outdated), we assume the total supply will be outdated as well and will
+            // have frozen tokens available in last/earlier period.
+        | None -> ((failwith "BAD_STATE") : nat) in
+      let previous_participation = { numerator = last_period_frozen_tokens; denominator = config.governance_total_supply } in
+      let old_quorum = store.quorum_threshold_at_stage.quorum_threshold in
+      let one = {numerator = 1n; denominator = 1n} in
+      let possible_new_quorum =
+        let term1 = mul_qt(old_quorum, sub_qt(one, {numerator = config.change_percent; denominator = 100n})) in
+        let term2 = mul_qt(previous_participation, {numerator = config.change_percent; denominator = 100n}) in
+        add_qt(term1, term2) in
+      let max_change_percent = { numerator = config.max_change_percent; denominator = 100n } in
+      let one_plus_max_change_percent = { numerator = 100n + config.max_change_percent; denominator = 100n } in
+      let min_new_quorum =
+        div_qt(old_quorum, one_plus_max_change_percent) in
+      let max_new_quorum =
+        mul_qt(old_quorum, one_plus_max_change_percent)
```

Comment:
I tried to reduce the cost of these calculations as much as possible,

Sadly it seems that using fraction-based operations adds a significant cost as probably ligo needs to pair and unpair many them continuously.

IMO we should go with a solution that takes as little as possible and leave comment to explain what's happening more clearly, here's what I could come up with:
```suggestion
      let old_quorum = store.quorum_threshold_at_stage.quorum_threshold in
      let old_quorum_numer = old_quorum.numerator in
      let old_quorum_denom = old_quorum.denominator in
      // participation = last_period_frozen_tokens / governance_total_supply
      // possible_new_quorum = old_quorum * (1 - change_percent) + participation * change_percent
      let possible_new_quorum =
        let term1 =
              { numerator = old_quorum_numer * sub_nat (100n, config.change_percent)
              ; denominator = old_quorum_denom * 100n
              }
        in
        let term2 =
              { numerator = last_period_frozen_tokens * config.change_percent
              ; denominator = config.governance_total_supply * 100n
              }
        in
        add_qt(term1, term2) in
      let one_plus_max_change_percent = 100n + config.max_change_percent in
      // min_new_quorum = old_quorum / (1 + max_change_percent)
      let min_new_quorum =
        { numerator = old_quorum_numer * 100n
        ; denominator = old_quorum_denom * one_plus_max_change_percent
        }
      in
      // max_new_quorum = old_quorum * (1 + max_change_percent)
      let max_new_quorum =
        { numerator = old_quorum_numer * one_plus_max_change_percent
        ; denominator = old_quorum_denom * 100n
        }
```
This cuts the size of the compiled contract quite a bit (`18982` -> `18818` bytes) and due to the lower number of operations I'd imagine the gas cost even more.

(note: you'll need a `sub_nat` which just throws an error if the subtraction is negative)

Reply:
> Sadly it seems that using fraction-based operations adds a significant cost as probably ligo needs to pair and unpair many them continuously.

Would it be better to replace numerator denominator pairs with a single `nat` with a known precision. AFAIU this is how Tezos core does it. For example `1.95` will be stored as `195`. 

Can it significantly reduce the number of calculations?

Reply:
@sras that's a very good idea, let's go with it! 

However I still think that percentage is not precise enough, I suggest once again that we use a per-million instead.

That should be safe as we know that the biggest numerator is a million as well and that the biggest number we can obtain within an operation is when multiplying two numbers, so `10 ^ 12`, which is still smaller than the max `nat` (that AFAIK is represented like a `63` bit unsigned int).

> same with division as well and a/b can be just `div (a*100) b`

Actually `a/b` should be just ... `a/b` too (the rest looks correct).

BTW I suggest documenting this properly and still keep a type synonym to better keep track of this:
```ocaml
// Quorum threshold that a proposal needs to meet in order to be accepted,
// expressed as a fraction of the total_supply of frozen tokens.
//
// For efficiency, we only keep a `nat` for the numerator, whereas the denominator
// is not stored and has a fixed value of `1000000`.
//
// Invariant: numerator < denominator
type quorum_threshold = nat
```

Reply:
>Actually a/b should be just ... a/b too (the rest looks correct).

But it seems that it would lose precision , because the decimal places will be lost. for ex `24.56/47.34` would be `2456/4734` which would be just `0`. But multiplying numerator with `100`, it would result in `51`, which the approximation we want in this representation of the actual answer, `0.5188`.

Agree with the rest.

Reply:
> it seems that it would lose precision

Oh yes, good point, I didn't think about the fact that it was a `nat` operation for a moment, never mind my comment there :+1:

Reply:
@pasqu4le I have included this change. In addition, I have also changed it to use `int` instead of `nat` so that it could momenerly accomodate negative values to simplify calculations. And consequently, I have changed some `config` values to be `int` so that it ll save some `nat`/`int` conversion in quorum_calculation.

Reply:
@sras instead of changing the type to be an `int`, which could store a potentially wrong value (because negative) I think we should keep the `nat` in `quorum_threshold`.

This should not be a problem as it's not necessary for us to hold a `nat`/`quorum_threshold` after every operation, we can use `int` in the middle steps and then make one conversion before saving the result.

Reply:
Have changed quorum threshold to be an Int.